Fix gcc-10 compilaton issues

michalpasztamobica opened the pull request:
When trying to build flang using gcc-10 on top of `classic-flang-llvm-project` I came across some issues which were overlooked by gcc-9:
* an uninitialized array in one of `libpgmath`'s files (trivial to fix),
* a number of global symbols defined in header files (resulting in multiple definitions errors).

Regarding the second issue there is at least one case where the error is about a generated variable `STB stb` which lives in `symtabdf.h`, written by a utility program `fesymutil`. 

In the easier case it is quite straightforward to modify the code, so it doesn't give us multiple definition errors with `-fno-common` by just naming the structures and providing `extern` symbols in the header files and then moving their declarations to one of the c files. This approach is presented in `prstab.h` and `prstab1.c`.

Before diving into the other issues (at least one of them coming from a generated file) I wanted to check if it would be acceptable to just add the `-fcommon` flag to workaround the definitions in header files? 

@RichBarton-Arm , @kiranchandramohan , your initial thoughts and comparison with Arm's downstream code would be much appreciated.

d-parks commented:
@michalpasztamobica , could you please send me the failure you're getting with gcc 10 (major.minor?) with libm_inlines_amd.h?  Also the gcc command line used? Thank you.

michalpasztamobica commented:
Hi @d-parks , thanks for reaching out!
I am running a docker image of Ubuntu-20.04 with gcc-10.2.0.
The command I used to build https://github.com/flang-compiler/classic-flang-llvm-project/commit/710e16efe81e89a67fcc7f313527750863ee0d41:
```
cmake -DLLVM_TARGETS_TO_BUILD=X86 -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=/usr/bin/gcc-10 -DCMAKE_CXX_COMPILER=/usr/bin/g++-10 -DLLVM_ENABLE_PROJECTS="clang;openmp" -DLLVM_ENABLE_CLASSIC_FLANG=ON ../llvm
make
sudo make install
```
Then I built libpgmath as present in https://github.com/flang-compiler/flang/commit/05b05c583152d05c4a3b758ea0966bfedc4123bc inside `runtime/libpgmath/build` directory with:
```
cmake -DLLVM_TARGETS_TO_BUILD=X86 -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=/usr/bin/gcc-10 -DCMAKE_CXX_COMPILER=/usr/bin/g++-10 ..
make
```
and got:
```
[ 87%] Building C object lib/x86_64/CMakeFiles/avx2.dir/sincosp.c.o
In file included from /home/flang/runtime/libpgmath/lib/x86_64/sincosp.c:50:
/home/flang/runtime/libpgmath/lib/x86_64/sincosp.c: In function '__mth_sincos':
/home/flang/runtime/libpgmath/lib/x86_64/libm_inlines_amd.h:2041:22: error: 'res[5]' may be used uninitialized in this function [-Werror=maybe-uninitialized]
 2041 |     nextbits = (~(res[i + 1]) & mask);
      |                  ~~~~^~~~~~~~
/home/flang/runtime/libpgmath/lib/x86_64/libm_inlines_amd.h:2050:14: error: 'res[5]' may be used uninitialized in this function [-Werror=maybe-uninitialized]
 2050 |     nextbits = res[i + 1];
      |     ~~~~~~~~~^~~~~~~~~~~~
/home/flang/runtime/libpgmath/lib/x86_64/libm_inlines_amd.h:2050:14: error: 'res[6]' may be used uninitialized in this function [-Werror=maybe-uninitialized]
/home/flang/runtime/libpgmath/lib/x86_64/libm_inlines_amd.h:2050:14: error: 'res[7]' may be used uninitialized in this function [-Werror=maybe-uninitialized]
/home/flang/runtime/libpgmath/lib/x86_64/libm_inlines_amd.h:2050:14: error: 'res[8]' may be used uninitialized in this function [-Werror=maybe-uninitialized]
/home/flang/runtime/libpgmath/lib/x86_64/libm_inlines_amd.h:2050:14: error: 'res[9]' may be used uninitialized in this function [-Werror=maybe-uninitialized]
```
The errors later repeaths in other files which include `libm_inlines_amd.h`

I ran make with `VERBOSE=1` getting:
```
cd /home/flang/runtime/libpgmath/build/lib/x86_64 && /usr/bin/gcc-10 -DFN_PROTO_SUFFIX=_avx2 -DHOST_LINUX -DLINUX -DLINUX86 -DLINUX8664 -DMAXCPUS=256 -DMAXCPUSL=8 -DMAXCPUSR=8 -DPG_PIC -DTARGET_LINUX -DTARGET_LINUX_X86 -DTARGET_LINUX_X8664 -DTARGET_X86 -DTARGET_X8664 -D_GNU_SOURCE -D_ISOC99_SOURCE -D__gnu_linux__ -I/home/flang/runtime/libpgmath/lib/common -I/home/flang/runtime/libpgmath/lib/x86_64 -O3 -DNDEBUG -fPIC -m64 -O3 -Wall -W -Wstrict-prototypes -Wwrite-strings -Werror  -mavx2 -mfma -ffp-contract=fast -o CMakeFiles/avx2.dir/sincosp.c.o -c /home/flang/runtime/libpgmath/lib/x86_64/sincosp.c
make  -f lib/x86_64/asin/avx/CMakeFiles/x86_64_asin_avx.dir/build.make lib/x86_64/asin/avx/CMakeFiles/x86_64_asin_avx.dir/depend
```

d-parks commented:
@michalpasztamobica Thank you.

gklimowicz reviewed (approved):
This passes our tests on OpenPOWER for LLVM 9 and LLVM 10.

michalpasztamobica commented:
Thanks for the review, @gklimowicz .
Would you also let me know what you think of this PR, @kiranchandramohan , @shivaramaarao , @bryanpkc ? Are you OK with using the `-fcommon` flag for now?

kiranchandramohan commented:
@michalpasztamobica I see -fcommon as well as changes to mark some globals as external. Not clear what (-fcommon, marking globals as external, renaming globals, or a mix of these) is exactly proposed in this patch.

michalpasztamobica commented:
Sorry about the confusion @kiranchandramohan .
The mix is the result of the fact that compilation issues show one at a time, to be honest. When I saw the first error I just fixed it properly (that is - I moved the global declarations to a .c file and used `extern` keyword in the header file, to allow access to the globals).
But when I saw the next failure, it was coming from a generated file. To fix it I would have to fix the generator and/or its source files, which I am afraid will consume some more time, while all I want is to be able to build with gcc-10 to enable extra compilers for flang.
Still, once I had one place fixed already I thought it's best to commit it rather than apply a workaround flag for that file.

So it's all about time - do we want a proper fix for all these issues (even if it takes a couple days) or can we leave the flags in for most of them and go on fixing other bugs?

michalpasztamobica commented:
There's also an extra fix in libpgmath, but that one is kept in a separate commit and I believe @d-parks is looking at it to judge if it makes sense or not?

kiranchandramohan commented:
@michalpasztamobica fno-common is the default for clang/llvm-11. The default building strategy is to build flang with clang/llvm. So this patch is important.

I think if it is two days work then i think it is worth to spend that time to fix it properly. However if it is two weeks then probably not. Another criteria could be whether it will be a few changes in files or big wholesale changes.
What do others (@bryanpkc @shivaramaarao @gklimowicz @mleair) think of the best way to handle this?

michalpasztamobica commented:
Hi @kiranchandramohan  . The fixes as such are pretty straightforward, but the generated files might be more of a challenge. I would however estimate for about 2-4 days in total.

bryanpkc commented:
I agree with @kiranchandramohan, if it's two to four days of work, we should spend that time to fix the problem properly. I'm not sure how many people or companies are using GCC 10 to build Flang today, but it is a good time to deal with this now rather than later when there is higher pressure to get GCC 10 working.

michalpasztamobica commented:
I just force pushed modified PR, which:
* still has the libpgmath fix (@d-parks , any further feedback on this?),
* still has the fixes for global symbols it had before,
* instead of `-fcommon` flag added to CMakeFiles we now have some more code fixes, including a change in the `fesymutil` tool, which involves interface change - an extra parameter to name the new `.c` file.

This last change requires special attention as there might be build methods and commands that I might not know about which will now require the extra parameter. 

This [is already tested](https://github.com/michalpasztamobica/flang/actions/runs/395827653) against all compilers (gcc-9/10, llvm-9/10/11) in the context of the dependent PR.

@Reviewers, please let me know any remarks to this.

d-parks commented:
@michalpasztamobica - go ahead with your the fix to libm_inlines_amd.h.  I have yet to determine whether the array is sized too large, or we really have a read before write condition.

michalpasztamobica commented:
@gklimowicz , @bryanpkc , @kiranchandramohan , kind request for you to take another look at this patch, now being a proper code fix, rather than a workaround.

Please, pay particular attention to the changes in the generator, where I had to add another command line option to the `fesymutil` tool, effectively changing the interface. If you know of any situations (like some non-standard build modes), which could be affected, please let me know.

gklimowicz reviewed (approved):
This passes our tests on OpenPOWER for LLVM 9 and LLVM 10.

I ran this with both GCC 10.2 and GCC 7.3

michalpasztamobica commented:
@bryanpkc , @kiranchandramohan , @shivaramaarao , would you please give you code reviews? Thank you in advance!

michalpasztamobica commented:
Hi @bryanpkc , @kiranchandramohan , @shivaramaarao , just a friendly reminder. Would you be able to give your reviews here? Thanks!

bryanpkc reviewed (approved):
LGTM.

shivaramaarao reviewed (approved):
Looks fine to me.

michalpasztamobica commented:
@kiranchandramohan , would you have a moment to review this patch after the additional changes and perhaps merge it if it's OK?

kiranchandramohan reviewed (commented):
Looks OK. I have a question.

michalpasztamobica commented on tools/shared/utils/symacc.c:
```diff
@@ -36,11 +36,11 @@
 #else
 extern STB stb;
-extern GBL gbl;
+GBL gbl;
```
These `gbl`s are different from the one I touched in this PR. Removing them causes an "undefined reference" error. But when they are in the code compiles, so they can't be overlapping with the `gbl` which I modified as that would cause a "multiple definition error" we have seen earlier.
Apparently - there are multiple structures, which are declared separately, perhaps on purpose. They have to stay as they are...

michalpasztamobica commented:
@kiranchandramohan , I addressed your first remark (`cpp` file extenstion rather than `c`) and hopefully answered the other one. Let me know if you see any more issues...

kiranchandramohan reviewed (approved):
LGTM. Checked with gcc-10 and clang-11.

@michalpasztamobica should this be squashed?

michalpasztamobica commented:
@kiranchandramohan , I would personally leave the commits separate (for example - in case of any future need for git bisection), as they are three distinct changes to me.
But if you prefer, I can squash, just let me know,
